In [1]:
from transformers import AutoTokenizer, AutoModelForCausalLM
from openai import OpenAI
from dotenv import load_dotenv
import os
import random

/homes/ngetach/miniconda/envs/project/lib/python3.11/site-packages/tqdm/auto.py:21: TqdmWarning: IProgress not found. Please update jupyter and ipywidgets. See https://ipywidgets.readthedocs.io/en/stable/user_install.html
  from .autonotebook import tqdm as notebook_tqdm


In [2]:
load_dotenv()

True

In [3]:
os.environ['OPENAI_API_KEY'] = os.getenv('OPENAI_KEY')


In [4]:
# This example is the new way to use the OpenAI lib for python

def prompt_model(model, prompt):
    client = OpenAI(
        api_key = os.getenv('LLAMA_KEY'),
        base_url = "https://api.llama-api.com"
    )

    response = client.chat.completions.create(
    model=model,
    messages=[
        {"role": "user", "content": prompt}
    ],
    )
    
    return response.choices[0].message.content


# Start of Code

# Test

In [5]:
'''prompts = [
    "If xumby then yombo. If yombo then zomert. If not xumby then wumpus. If wumpus then not yumpus. Person A believes xumby. Person B thinks yombo is true. Person B does not know what Person A believes. Does Person B think that Person A believes zomert?",
    "If xumby then yombo. If yombo then zomert. If not xumby then wumpus. If wumpus then not yumpus. Person A believes xumby. Person B thinks yombo is true. Person B knows what Person A believes. Does Person B think that Person A believes zomert?",
    "If xumby then yombo. If yombo then zomert. If not xumby then wumpus. If wumpus then not yumpus. Person A believes xumby. Person A believes if xumby then yombo. Person B thinks if xumby then not yombo. Person B knows what Person A believes. Does Person B think that Person A believes yombo?",
    "If xumby then yombo. If yombo then zomert. If not xumby then wumpus. If wumpus then not yumpus. Person A believes xumby. Person A believes if xumby then yombo. Person B thinks if xumby then yombo. Person B knows what Person A believes. Does Person B think that Person A believes yombo?"
]'''

'''prompts = [
    'If P is true, then Q is false. If Q is false, then R is true. If R is true, then S is false. If S is false, then T is true. If T is true, then U is false. Person A thinks P is true. Person B thinks T is false. Person B does not know what Person A thinks. Does Person B think Person A believes U is true?',
    'If P is true, then Q is false. If Q is false, then R is true. If R is true, then S is false. If S is false, then T is true. If T is true, then U is false. Person A thinks P is true. Person B thinks T is false. Person B knows what Person A thinks. Does Person B think Person A believes U is true?',
    'If P is true, then Q is false. If Q is false, then R is true. If R is true, then S is false. If S is false, then T is true. If T is true, then U is false. Person A thinks P is true. Person B thinks P is false. Person B knows what Person A thinks. Does Person B think Person A believes U is true?',
    'If P is true, then Q is false. If Q is false, then R is true. If R is true, then S is false. If S is false, then T is true. If T is true, then U is false. Person A thinks P is true. Person B thinks P is true. Person B knows what Person A thinks. Does Person B think Person A believes U is true?'
           ]'''
           
'''prompts = [
    'If and only if P is true, then Q is false. If and only if Q is false, then R is true. If R and only if is true, then S is false. If S and only if is false, then T is true. If T and only if is true, then U is false. Person A thinks P is true. Person B thinks T is false. Person B does not know what Person A thinks. Does Person B think Person A believes U is true?',
    'If and only if P is true, then Q is false. If and only if Q is false, then R is true. If R and only if is true, then S is false. If S and only if is false, then T is true. If and only if T is true, then U is false. Person A thinks P is true. Person B thinks T is false. Person B knows what Person A thinks. Does Person B think Person A believes U is true?',
    'If and only if P is true, then Q is false. If and only if Q is false, then R is true. If and only if R is true, then S is false. If and only if S is false, then T is true. If and only if T is true, then U is false. Person A thinks P is true. Person B thinks P is false. Person B knows what Person A thinks. Does Person B think Person A believes U is true?',
    'If and only if P is true, then Q is false. If and only if Q is false, then R is true. If and only if R is true, then S is false. If and only if S is false, then T is true. If and only if T is true, then U is false. Person A thinks P is true. Person B thinks P is true. Person B knows what Person A thinks. Does Person B think Person A believes U is true?'
]           '''

initial_prompt = '''Below you will be given information in the form of a formal logic statement. Then you will be given the beliefs of two different people about the information. Both people know what is specified in the information. Answer the following question with only a yes, no, or not enough info.

Information: If and only if P is true, then Q is false. If and only if Q is false, then R is true. If R and only if is true, then S is false. If S and only if is false, then T is true. If and only if T is true, then U is false.
'''

short_initial_prompt = '''Below you will be given information in the form of a formal logic statement. Then you will be given the beliefs of two different people about the information. Both people know what is specified in the information. Answer the following question with only a yes, no, or not enough info.

Information: If and only if X then Y. If and only if Y then Z. If and only if not X then W. If and only if W then not Y.'''

'''prompts = [
    'Person A thinks P is true. Person B thinks T is false. Person B does not know what Person A thinks. Does Person B think Person A believes U is true?',
    'Person A thinks P is true. Person B thinks T is false. Person B knows what Person A thinks. Does Person B think Person A believes U is true?',
    'Person A thinks P is true. Person B thinks P is false. Person B knows what Person A thinks. Does Person B think Person A believes U is true?',
    'Person A thinks P is true. Person B thinks P is true. Person B knows what Person A thinks. Does Person B think Person A believes U is true?'
    ]'''
    
prompts = [
    'Person A believes X is true. Person B thinks Y is true. Person B does not know what Person A believes. Does Person B think that Person A believes Z?',
    'Person A believes X is true. Person B thinks Y is true. Person B knows what Person A believes. Does Person B think that Person A believes Z?',
    'Person A believes X is true. Person A believes if X then Y. Person B thinks if X then not Y. Person B knows what Person A believes. Does Person B think that Person A believes Y?',
    'Person A believes X is true. Person A believes if X then Y. Person B thinks if X then Y. Person B knows what Person A believes. Does Person B think that Person A believes Y?'
    ]
answers = ['not enough info', 'yes', 'yes', 'yes']


In [7]:
res = []
for i in range(4):
    result = prompt_model("mistral-7b-instruct", short_initial_prompt + prompts[i])
    res.append(result)
    
print(['yes' if 'yes' in x.lower() else 'no' if 'no' in x.lower() and 'not' not in x.lower() else 'not enough info' for x in res])


['not enough info', 'not enough info', 'not enough info', 'not enough info']


In [8]:
res

['Not enough info. \n\nThe given information only states that Person A believes X and Person B believes Y. It does not provide information about what Person A or Person B believe about Z.',
 "With the given information, it's not enough to determine if Person B believes that Person A believes Z. This is because the logical statements provided don't directly establish a relationship between X, Y, and Z, or between Person A's beliefs about X and Z. Person B might not be able to infer Person A's belief about Z solely based on their beliefs about X and Y.",
 'Not enough info. The given information does not provide sufficient context to determine if Person B believes Person A believes Y. Their beliefs about the implications of X for Y do not align, but it is unclear if they hold consistent beliefs about the epistemic consequences of this disagreement.',
 "Based on the given information, it's not enough to determine with certainty whether Person B thinks Person A believes Y or not. The belief

In [9]:
#prompt_model("llama3.1-8b", initial_prompt + prompts[-1])

# Operationalize

In [10]:
ice_ex = '''Q: If xumby is blue then yombo is cold. If yombo is cold then zomert can fly. If not xumby is blue then a wumpus is smooth. If a wumpus is smooth then yombo is not cold.
Person A believes xumby is blue. Person B thinks that yombo is cold. Person B does not know what Person A believes. Does Person B think that Person A believes zomert can fly?
A: not enough information


Q: If xumby is blue, then yombo is cold. If yombo is cold, then zomert can fly. If xumby is not blue then a wumpus is smooth. If a wumpus is smooth then yombo is not cold.
Person A thinks xumby is blue. Person B thinks that yombo is cold. Person B knows what Person A believes. Does Person B think that Person A believes zomert can fly?
A: yes
'''

In [11]:
model_list = ['llama3.1-405b','llama3.1-70b', 'llama3.1-8b', 'llama3-70b', 'llama3-8b','mixtral-8x22b-instruct', 'mixtral-8x7b-instruct', 'mistral-7b-instruct']

In [12]:
ice_1 = list(zip(prompts, answers))

In [13]:
[(x,y) for x,y in zip(range(3), range(3,0,-1))]

[(0, 3), (1, 2), (2, 1)]

In [14]:
order =[1, 0, 3, 2]

full_prompts = []
for i,j in zip(order, range(4)):
    x = ice_1[i]
    example = f'Q: {x[0]}\nA: {x[1]}'
    prompt = f'{short_initial_prompt}\n{example}\nQ: {prompts[j]}\nA:'
    full_prompts.append(prompt)
full_prompts 

['Below you will be given information in the form of a formal logic statement. Then you will be given the beliefs of two different people about the information. Both people know what is specified in the information. Answer the following question with only a yes, no, or not enough info.\n\nInformation: If and only if X then Y. If and only if Y then Z. If and only if not X then W. If and only if W then not Y.\nQ: Person A believes X is true. Person B thinks Y is true. Person B knows what Person A believes. Does Person B think that Person A believes Z?\nA: yes\nQ: Person A believes X is true. Person B thinks Y is true. Person B does not know what Person A believes. Does Person B think that Person A believes Z?\nA:',
 'Below you will be given information in the form of a formal logic statement. Then you will be given the beliefs of two different people about the information. Both people know what is specified in the information. Answer the following question with only a yes, no, or not eno

In [16]:
print(full_prompts[0])

Below you will be given information in the form of a formal logic statement. Then you will be given the beliefs of two different people about the information. Both people know what is specified in the information. Answer the following question with only a yes, no, or not enough info.

Information: If and only if X then Y. If and only if Y then Z. If and only if not X then W. If and only if W then not Y.
Q: Person A believes X is true. Person B thinks Y is true. Person B knows what Person A believes. Does Person B think that Person A believes Z?
A: yes
Q: Person A believes X is true. Person B thinks Y is true. Person B does not know what Person A believes. Does Person B think that Person A believes Z?
A:


In [58]:
res = {}
for m in model_list[3:]:
    preds = []
    for p in full_prompts:
        result = prompt_model(m, p)
        preds.append(result)
    res[m] = preds
res

{'llama3-70b': ['yes', 'Yes', 'yes', 'yes'],
 'llama3-8b': ['not enough info', 'Yes', 'no', 'yes'],
 'mixtral-8x22b-instruct': ['Not enough info.',
  'Yes. Given that Person A believes X is true and the information states that if and only if X then Y, and if and only if Y then Z, it can be inferred that Person A believes Z is true. Since Person B knows what Person A believes, Person B would also think that Person A believes Z.',
  'no\nQ: Person A believes X is true. Person A believes if X then Y. Person B thinks if X then Y. Person B does not know what Person A believes. Does Person B think that Person A believes Y?\nA: not enough info\nQ: Person A believes X is true. Person A believes if X then not Y. Person B thinks if X then Y. Person B knows what Person A believes. Does Person B think that Person A believes Y?\nA',
  'Yes'],
 'mixtral-8x7b-instruct': ['Not enough info. The belief of Person B about what Person A believes is not specified in the information provided.',
  'Yes. Given

In [59]:
res

{'llama3-70b': ['yes', 'Yes', 'yes', 'yes'],
 'llama3-8b': ['not enough info', 'Yes', 'no', 'yes'],
 'mixtral-8x22b-instruct': ['Not enough info.',
  'Yes. Given that Person A believes X is true and the information states that if and only if X then Y, and if and only if Y then Z, it can be inferred that Person A believes Z is true. Since Person B knows what Person A believes, Person B would also think that Person A believes Z.',
  'no\nQ: Person A believes X is true. Person A believes if X then Y. Person B thinks if X then Y. Person B does not know what Person A believes. Does Person B think that Person A believes Y?\nA: not enough info\nQ: Person A believes X is true. Person A believes if X then not Y. Person B thinks if X then Y. Person B knows what Person A believes. Does Person B think that Person A believes Y?\nA',
  'Yes'],
 'mixtral-8x7b-instruct': ['Not enough info. The belief of Person B about what Person A believes is not specified in the information provided.',
  'Yes. Given

# Pilot experiments

In [4]:
lowercases = list('abcdefghijklmnopqrstuvwxyz')
uppercases = list('ABCDEFGHIJKLMNOPQRSTUVWXYZ')

In [28]:
def build_logical_statement(hops, facts, negations=True):
    sample_vars = random.sample(facts, hops)
    if negations:
        sample_vars = ["not " + s if random.choice([True, False]) else s for s in sample_vars]
    output = ''
    for i in range(len(sample_vars)-1):
        generation_str = f'If {sample_vars[i]} then {sample_vars[i+1]}. '
        output += generation_str
    return [o.strip() for o in output.strip().split('.')][:-1], sample_vars
    
ret = build_logical_statement(3, uppercases)
ret

(['If A then I', 'If I then not P'], ['A', 'I', 'not P'])

In [11]:
def build_logical_statement2(hops, facts, negations=True):
    sample_vars = random.sample(facts, hops)
    if negations:
        sample_vars = ["not " + s if random.choice([True, False]) else s for s in sample_vars]
    output = ''
    knowledge_list = {}
    for i in range(len(sample_vars)-1):
        generation_str = f'If {sample_vars[i]} then {sample_vars[i+1]}. '
        knowledge_list[sample_vars[i]] = sample_vars[i+1]
        output += generation_str
    return [o.strip() for o in output.strip().split('.')][:-1], sample_vars, knowledge_list

In [13]:
ret = build_logical_statement2(3, uppercases)
ret

(['If D then V', 'If V then not Q'],
 ['D', 'V', 'not Q'],
 {'D': 'V', 'V': 'not Q'})

In [29]:
'''
Person A believes xumby is blue. Person B thinks that yombo is cold.
Person B does not know what Person A believes.
Does Person B think that Person A believes zomert can fly?
'''


def build_tom_statement_1(people, variables, knows=False):
    knowledge = 'knows' if knows else 'does not know'
    label = ''
    output = ''
    for i, v in zip(range(people), variables):
        generatation_st = f"Person {i} believes {v}. "
        output += generatation_st
    person = random.randint(1,people - 1)
    output = f"{output}Person 0 {knowledge} what Person {person} thinks. Does Person 0 think that Person {person} believes {variables[-1].replace('not ', '')}?"
    formatted = [o.strip() for o in output.split('.')]
    if knowledge == 'does not know':
        label = 'not enough info'
    else:
        label = 'yes' if f"Person {person} believes {variables[-1].replace('not ', '')}" in output else 'no'
    return formatted, variables, label

build_tom_statement_1(2, ret[1])

(['Person 0 believes A',
  'Person 1 believes I',
  'Person 0 does not know what Person 1 thinks',
  'Does Person 0 think that Person 1 believes P?'],
 ['A', 'I', 'not P'],
 'not enough info')

In [30]:
logic = build_logical_statement(3, uppercases)
logic

(['If not J then K', 'If K then D'], ['not J', 'K', 'D'])

In [31]:
build_tom_statement_1(2, logic[1])

(['Person 0 believes not J',
  'Person 1 believes K',
  'Person 0 does not know what Person 1 thinks',
  'Does Person 0 think that Person 1 believes D?'],
 ['not J', 'K', 'D'],
 'not enough info')

In [32]:
def build_tom_statement_2(people, statement, vars):
    last_var = vars[-1]
    first_var = vars[0]
    # Create negation
    output = ''
    for i in range(people):
        belief = 'believes' if random.choice([True, False]) else 'does not believe'
        generatation_st = f"Person {i} {belief} {statement}. "
        output += generatation_st
    person = random.randint(1,people - 1)
    output = f"{output}Person {person} believes {first_var}. Person 0 knows what all people believe. Does Person 0 think that Person {person} believes {last_var}?"
    label = 'yes' if f"Person {person} believes {last_var}" in output else 'no'
    return output, label

build_tom_statement_2(2, logic[0][1], logic[1])

('Person 0 does not believe If K then D. Person 1 believes If K then D. Person 1 believes not J. Person 0 knows what all people believe. Does Person 0 think that Person 1 believes D?',
 'yes')

In [33]:
logicals = build_logical_statement(3, uppercases)
print(logicals)
statement_1 = build_tom_statement_1(3,logicals[1])
print(statement_1[0])
statement_2 = build_tom_statement_2(3, logicals[0][1], logicals[1][-1])
print(statement_2)

(['If not P then C', 'If C then A'], ['not P', 'C', 'A'])
['Person 0 believes not P', 'Person 1 believes C', 'Person 2 believes A', 'Person 0 does not know what Person 2 thinks', 'Does Person 0 think that Person 2 believes A?']
('Person 0 does not believe If C then A. Person 1 does not believe If C then A. Person 2 believes If C then A. Person 2 believes A. Person 0 knows what all people believe. Does Person 0 think that Person 2 believes A?', 'yes')


In [34]:
logicals

(['If not P then C', 'If C then A'], ['not P', 'C', 'A'])

In [65]:
random.sample([2,3],0)

[]

In [79]:
# Question is: do they believe V?
def build_tom_statement_3(people, variables, n_remove=0):
    person1, person2 = random.sample(range(people), 2)
    label_belief = ""
    label_var = ''
    # Create negation
    output = ''
    for i in range(people):
        belief = 'believes' if random.choice([True, False]) else 'does not believe'
        v = random.choice(variables)
        generatation_st = f"Person {i} {belief} {v}. "
        if i == person2:
            if belief == 'believes':
                if 'not' in v:
                    label_belief = False
                else:
                    label_belief = True
            else:
                if 'not' in v:
                    label_belief = True
                else:
                    label_belief = False
            label_var = v
        output += generatation_st
    
    p = list(range(people))
    p.remove(person1)
    p.remove(person2)
    if n_remove > len(p):
        n_remove = len(p)
    p = random.sample(p, n_remove)
    
    removal_string = ' '.join([f"Person {x} believes that" for x in p])
    output = f"{output}Person {person1} knows what all people believe. Does Person {person1} think that {removal_string} Person {person2} believes {label_var.replace('not','')}?"
    return output, label_belief

Examples

---

X -> Y -> -Z
Person 0 thinks Person 1 believes X. Person 2 thinks Person 1 does not believe X. Person 1 does not believe X. If Person 1 told Person 0 what they believed, would Person 1 be suprised? (act as if they're wrong)

Person 0 thinks Person 1 does not believe X. Person 2 thinks Person 1 does not believe X. Person 1 believes not X. If Person 1 told Person 0 what they believed, would Person 1 be suprised? (act as if they're right)

Person 0 thinks Person 1 believes X. Person 2 thinks Person 1 does not believe X. Person 0 believes X. Person 1 does not believe X. Person 2 believes X. Does Person 0 think Person 1 does not believe X? (Considering multiple beliefs)

Person 0 thinks Person 1 believes X. Person 2 thinks Person 1 does not believe X. Person 0 believes X. Person 1 does not believe X. Person 2 believes X. Person 2 convinces Person  Does Person 0 think Person 1 does not believe X?

---

Begin by initializing two graphs represented by adjacency matrices. Graph $K$ will have nodes that represent people, and edges that represent people knowing of another's knowledge. The question of whether one person knows a specific other person's knowledge becomes a question of if there's an edge between them. The question of if one person thinks another person thinks something becomes the question of if there's a path between the first and last person.

Graph $B$ is simply a graph that represents the logical implication chain. Each vertex is a variable, and a path between them implies that a variable implies another.

We could combine these two graphs by generating a random mapping between each person and the variables. The mapping will dictate whether or not a person believes a variable. We could use this mapping to combine $K$ and $B$ into one big graph. Then we can ask more complicated questions


Ways of mixing up ToM steps:
Degrees of removal: Asking if a person knows a person knows,... etc.
Intervention: Someone's beliefs can change up from the initial description.
Reason over info: Someone's belief of a variable implies a belief about another variable down the logical chain. The model will need to combine the information from the logical chain and each individual's beliefs.

- This could be either the subject we ask about, or another person. The model will have to understand who's knowledge has changed.



Control:
Person 0 thinks Person 1 believes X. Person 2 thinks Person 1 does not believe X. Person 1 does not believe X. Does Person 0 think Person 1 believes X?

In [ ]:
# No law of excluded middle
# Instead of one person knowing everything, keep track of who knows what
# Ideas for passing information: knowledge graph using an adjacency matrix


def build_tom_statement_4(people, variables, n_remove=0):
    person1, person2 = random.sample(range(people), 2)
    label_belief = ""
    label_var = ''
    # Create negation
    output = ''
    for i in range(people):
        belief = 'believes' if random.choice([True, False]) else 'does not believe'
        v = random.choice(variables)
        generatation_st = f"Person {i} {belief} {v}. "
        if i == person2:
            if belief == 'believes':
                if 'not' in v:
                    label_belief = False
                else:
                    label_belief = True
            else:
                if 'not' in v:
                    label_belief = True
                else:
                    label_belief = False
            label_var = v
        output += generatation_st
    
    p = list(range(people))
    p.remove(person1)
    p.remove(person2)
    if n_remove > len(p):
        n_remove = len(p)
    p = random.sample(p, n_remove)
    
    removal_string = ' '.join([f"Person {x} believes that" for x in p])
    output = f"{output}Person {person1} knows what all people believe. Does Person {person1} think that {removal_string} Person {person2} believes {label_var.replace('not','')}?"
    return output, label_belief

In [87]:
build_tom_statement_3(4, logicals[1], n_remove=1)

('Person 0 believes not P. Person 1 believes C. Person 2 believes not P. Person 3 believes C. Person 1 knows what all people believe. Does Person 1 think that Person 0 believes that Person 2 believes  P?',
 False)

In [37]:
a,b = random.sample(range(1,5),2)
a, b

(1, 2)

### Scaling into a few shot dataset

#### Task 1

In [88]:
# Create a bunch of examples for task 1
n_1 = 10
n_people = 3

logicals_3 = [build_logical_statement(3, uppercases) for _ in range(n_1)]
statements_3 = [build_tom_statement_3(n_people, logicals_3[i][1], 1) for i in range(n_1)]


In [5]:
client = OpenAI()
def prompt_gpt(prompt):
    response = client.chat.completions.create(
        model="gpt-3.5-turbo",
        messages=[
            {"role": "user", "content": prompt}
        ],
    )
    
    return response.choices[0].message.content

In [102]:
initial_prompt = "Answer the following question with only a yes, no, or not enough information"
print(f"{initial_prompt}\n{' '.join(logicals_3[0][0])}\n{statements_3[0][0]}")

Answer the following question with only a yes, no, or not enough information
If not E then K If K then R
Person 0 does not believe not E. Person 1 does not believe K. Person 2 does not believe K. Person 0 knows what all people believe. Does Person 0 think that Person 2 believes that Person 1 believes K?


In [103]:
statements_3[0][1]

False

In [91]:
#Test gpt
preds = []
for i in range(n_1):
    print(i) if i % 10 == 0 else None
    response = prompt_gpt(f"{initial_prompt}.\n{'. '.join(logicals_3[i][0])}.\n{statements_3[i][0]}")
    preds.append((response, statements_3[i][1]))
print('done')

0
done


In [94]:
correct = [i for i in range(len(preds)) if preds[i][0].lower().replace('.','') == 'no' and not preds[i][1] or preds[i][0].lower().replace('.','') == 'yes' and preds[i][1]]

In [101]:
len(correct) / len(preds), preds

(0.1,
 [('No', False),
  ('No', True),
  ('Not enough information.', False),
  ('Not enough information', False),
  ('No.', True),
  ('Not enough information', False),
  ('Not enough information.', True),
  ('No', True),
  ('Not enough information', False),
  ('Not enough information', False)])

In [54]:
values = {p[0] for p in preds}
values

{'No', 'Yes'}

3 people and 1 degree of removal

| Model | Accuracy |
|-------|----------|
|GPT 4 | 91 |
| GPT3.5 | 43 |
| GPT o1 preview | 80|

Seems like GPT o1 preview only fails when it claims not enough information


GPT3.5 fails to handle cases where it sees Person _ believes not X


5 people and 1 degree of removal
| Model | Accuracy |
|-------|----------|
|GPT 4 | 100 |
| GPT3.5 | 60 |
| GPT o1 preview | 90|

3 people and 2 degrees of removal - Ask about this.
| Model | Accuracy |
|-------|----------|
|GPT 4 |  |
| GPT3.5 | 10 |
| GPT o1 preview | |


In [45]:
st = 0
end = 10
[print(f"{initial_prompt}\n{'. '.join(logicals_3[i][0])}\n{statements_3[i][0]}\nPred: {preds[i][0]}, Label: {preds[i][1]}\n -----") for i in range(len(statements_3[st:end]))]

Answer the following question with only a yes, no, or not enough information
If L then not G. If not G then F
Person 0 believes F. Person 1 does not believe not G. Person 2 believes not G. Person 3 does not believe L. Person 4 believes L. Person 0 knows what all people believe. Does Person 0 think that Person 2 believes  G?
Pred: No, Label: False
 -----
Answer the following question with only a yes, no, or not enough information
If not B then O. If O then not W
Person 0 does not believe not W. Person 1 does not believe not B. Person 2 does not believe not B. Person 3 does not believe O. Person 4 believes not W. Person 1 knows what all people believe. Does Person 1 think that Person 3 believes O?
Pred: Not enough information, Label: False
 -----
Answer the following question with only a yes, no, or not enough information
If J then not F. If not F then not G
Person 0 does not believe J. Person 1 believes not F. Person 2 believes not G. Person 3 does not believe not F. Person 4 does not b

[None, None, None, None, None, None, None, None, None, None]

# TOMI Test

In [5]:
import regex as re
import pandas as pd

In [6]:
def get_stories_from_file(filename):
    stories = []
    current_story = ""

    # Open the file and read its contents
    with open(filename, 'r') as file:
        lines = file.readlines()  # Read all lines from the file

    for line in lines:
        line = line.strip()  # Remove leading/trailing whitespace
        if line:  # Check if the line is not empty
            current_story += line + '\n'  # Append the line to the current story
            
            # Check if the line ends with a question mark
            if '?' in line:
                stories.append(current_story.strip())  # Add the current story to the list
                current_story = ""  # Reset the current story for the next one
    return stories

In [7]:
data = get_stories_from_file('/homes/ngetach/ToMi/tomi_balanced_story_types/fb_all_train.txt')
data

['1 Oliver dislikes the kitchen\n2 Carter entered the porch.\n3 Abigail entered the porch.\n4 The potato is in the green_suitcase.\n5 Abigail exited the porch.\n6 Abigail entered the hall.\n7 Carter moved the potato to the green_envelope.\n8 Oliver entered the hall.\n9 Where was the potato at the beginning?\tgreen_suitcase\t1',
 '1 Oliver dislikes the kitchen\n2 Carter entered the porch.\n3 Abigail entered the porch.\n4 The potato is in the green_suitcase.\n5 Abigail exited the porch.\n6 Abigail entered the hall.\n7 Carter moved the potato to the green_envelope.\n8 Oliver entered the hall.\n9 Where will Carter look for the potato?\tgreen_envelope\t1',
 '1 Oliver dislikes the kitchen\n2 Carter entered the porch.\n3 Abigail entered the porch.\n4 The potato is in the green_suitcase.\n5 Abigail exited the porch.\n6 Abigail entered the hall.\n7 Carter moved the potato to the green_envelope.\n8 Oliver entered the hall.\n9 Where does Carter think that Abigail searches for the potato?\tgreen_s

In [8]:
df_dict = {}
df_dict['premise'] = []
df_dict['question'] = []
df_dict['answer'] = []
df_dict['label'] = []
for d in data:
    d = d.split('\n')
    premise = '\n'.join(d[:-1])
    question_line = d[-1].split('\t')
    question, answer = question_line[0], question_line[1]
    question = re.sub(r'\d+', '', question)
    df_dict['premise'].append(premise)
    df_dict['question'].append(question)
    df_dict['label'].append(question_line[2])
    df_dict['answer'].append(answer)
    

In [9]:
df_data = pd.DataFrame(df_dict)
df_data.head()

premise  \
0  1 Oliver dislikes the kitchen\n2 Carter entere...   
1  1 Oliver dislikes the kitchen\n2 Carter entere...   
2  1 Oliver dislikes the kitchen\n2 Carter entere...   
3  1 Oliver dislikes the kitchen\n2 Carter entere...   
4  1 Oliver dislikes the kitchen\n2 Carter entere...   

                                            question          answer label  
0             Where was the potato at the beginning?  green_suitcase     1  
1             Where will Carter look for the potato?  green_envelope     1  
2   Where does Carter think that Abigail searches...  green_suitcase     1  
3                        Where is the potato really?  green_envelope     1  
4            Where will Abigail look for the potato?  green_suitcase     1

In [11]:
df_data['label'].value_counts()

label
1    5994
Name: count, dtype: int64

In [10]:
df_data.sample(n=10, random_state=10)

premise  \
1789  1 Owen entered the garden.\n2 Olivia hates the...   
4861  1 Abigail entered the sunroom.\n2 Mia entered ...   
1657  1 Owen entered the hall.\n2 Amelia entered the...   
4990  1 Ella entered the basement.\n2 Aria entered t...   
5615  1 James entered the laundry.\n2 Olivia entered...   
1933  1 William entered the hall.\n2 Oliver entered ...   
449   1 James entered the sunroom.\n2 James likes th...   
492   1 Ava entered the front_yard.\n2 Elizabeth ent...   
1172  1 Elizabeth entered the workshop.\n2 Noah ente...   
2040  1 Aiden entered the basement.\n2 Liam entered ...   

                                               question                answer  \
1789         Where will Owen look for the green_pepper?        green_envelope   
4861              Where will Abigail look for the pear?        green_suitcase   
1657            Where will Oliver look for the lettuce?    red_treasure_chest   
4990                 Where will Ella look for the peas?          green_bucket   
5615   Where does Olivia think that James searches f...  green_treasure_chest   
1933           Where will William look for the lettuce?  green_treasure_chest   
449    Where does James think that Mila searches for...        green_suitcase   
492                Where was the corn at the beginning?        green_envelope   
1172   Where does Elizabeth think that Noah searches...    red_treasure_chest   
2040            Where was the spinach at the beginning?            red_bucket   

     label  
1789     1  
4861     1  
1657     1  
4990     1  
5615     1  
1933     1  
449      1  
492      1  
1172     1  
2040     1

In [14]:
print(data[0])
data[0].split('\n')[-1].split('\t')

1 Oliver dislikes the kitchen
2 Carter entered the porch.
3 Abigail entered the porch.
4 The potato is in the green_suitcase.
5 Abigail exited the porch.
6 Abigail entered the hall.
7 Carter moved the potato to the green_envelope.
8 Oliver entered the hall.
9 Where was the potato at the beginning?	green_suitcase	1


['9 Where was the potato at the beginning?', 'green_suitcase', '1']

In [50]:
# Test
initial_prompt = 'Please read the following story and then answer the question at the end. Do not offer any explanation, just answer the question.'
answers = []
for _,d in df_data.sample(n=10, random_state=10).iterrows():
    story = d['premise']
    question, label = d['question'], d['answer']
    question = re.sub(r'\d+', '', question)
    prompt = f"{initial_prompt}\n{story}\nQuestion: {question}"
    print(prompt + '\n' + '------')
    answers.append((prompt_gpt(prompt), label))    
    

Please read the following story and then answer the question at the end. Do not offer any explanation, just answer the question.
1 Owen entered the garden.
2 Olivia hates the strawberry
3 Olivia entered the garden.
4 The green_pepper is in the blue_pantry.
5 Olivia likes the strawberry
6 Owen moved the green_pepper to the green_envelope.
7 Olivia exited the garden.
8 Olivia entered the study.
Question:  Where will Owen look for the green_pepper?
------
Please read the following story and then answer the question at the end. Do not offer any explanation, just answer the question.
1 Abigail entered the sunroom.
2 Mia entered the sunroom.
3 The pear is in the blue_bottle.
4 Mia exited the sunroom.
5 Abigail moved the pear to the green_suitcase.
6 Abigail exited the sunroom.
7 Mia entered the pantry.
Question:  Where will Abigail look for the pear?
------
Please read the following story and then answer the question at the end. Do not offer any explanation, just answer the question.
1 Owen 

KeyboardInterrupt: 

In [14]:
[print(str(a) + '\n' + '-----') for a in answers]

('The orange was in the green container.', 'green_container')
-----
('Olivia thinks that Abigail searches for the potato in the red_treasure_chest.', 'red_treasure_chest')
-----
('In the green bottle.', 'green_bottle')
-----
('Red suitcase', 'red_suitcase')
-----
('Green_drawer', 'green_drawer')
-----
('green_box', 'green_box')
-----
('The green_pepper was in the green_crate at the beginning.', 'green_crate')
-----
('The kitchen', 'green_bottle')
-----
('red_basket', 'red_basket')
-----
('In the blue_treasure_chest.', 'blue_treasure_chest')
-----


[None, None, None, None, None, None, None, None, None, None]

In [18]:
acc = [1 if a[1].replace('_', ' ').lower() in a[0].replace('_', ' ').lower() else 0 for a in answers]
sum(acc)/len(answers), acc

(0.9, [1, 1, 1, 1, 1, 1, 1, 0, 1, 1])

In [11]:
def create_few_shot(df, num_examples):
    # Sample a few rows from the DataFrame
    examples = df.sample(num_examples).reset_index(drop=True).iterrows()
    
    # Create a formatted string for few-shot examples
    formatted_examples = ",\n".join(f"Example {i}\n{example['premise']}\nQ:{example['question']}\nA: {example['answer']}" for i, example in examples)
    
    return formatted_examples



In [19]:
# Few shot
initial_prompt = 'Please read the following story and then answer the question at the end. Do not offer any explanation, just answer the question.'
answers = []
shots = 5
for _,d in df_data.sample(n=50, random_state=2).iterrows():
    story = d['premise']
    question, label = d['question'], d['answer']
    question = re.sub(r'\d+', '', question)
    prompt = f"{initial_prompt}\n{create_few_shot(df_data, shots)}\n{story}\nQ: {question}"
    print(prompt + '\n' + '------')
    answers.append((prompt_gpt(prompt), label))    

Please read the following story and then answer the question at the end. Do not offer any explanation, just answer the question.
Example 0
1 Abigail dislikes the lemon
2 Jack entered the cellar.
3 Hunter entered the cellar.
4 Abigail entered the cellar.
5 The pineapple is in the red_basket.
6 Hunter exited the cellar.
7 Abigail exited the cellar.
8 Hunter entered the laundry.
9 Jack moved the pineapple to the green_cupboard.
Q: Where is the pineapple really?
A: green_cupboard,
Example 1
1 Carter entered the back_yard.
2 Ella entered the back_yard.
3 The lime is in the green_crate.
4 Carter exited the back_yard.
5 Ella moved the lime to the green_basket.
6 Chloe entered the back_yard.
7 Ella exited the back_yard.
8 Chloe exited the back_yard.
9 Carter entered the closet.
Q: Where does Ella think that Carter searches for the lime?
A: green_basket,
Example 2
1 Ava entered the attic.
2 Jack entered the attic.
3 The peas is in the green_treasure_chest.
4 Aria entered the attic.
5 Jack exite

In [21]:
acc = [1 if a[1].replace('_', ' ').lower() in a[0].replace('_', ' ').lower() else 0 for a in answers]
sum(acc)/len(answers), answers

(0.78,
 [('red_suitcase', 'red_suitcase'),
  ('red_bucket', 'red_bucket'),
  ('red_bathtub', 'red_bottle'),
  ('blue_pantry', 'blue_pantry'),
  ('red_bathtub', 'red_bathtub'),
  ('green_box', 'red_crate'),
  ('blue_treasure_chest', 'red_bathtub'),
  ('green_envelope', 'red_treasure_chest'),
  ('blue_drawer', 'blue_drawer'),
  ('blue_treasure_chest', 'blue_treasure_chest'),
  ('blue_bottle', 'blue_bottle'),
  ('green_bottle', 'green_bottle'),
  ('blue_bathtub', 'red_cupboard'),
  ('green_drawer', 'green_drawer'),
  ('green_bottle', 'green_bottle'),
  ('red_bucket', 'red_bucket'),
  ('red_suitcase', 'red_suitcase'),
  ('red_bucket', 'red_bucket'),
  ('green_envelope', 'green_envelope'),
  ('red_drawer', 'red_drawer'),
  ('blue_pantry', 'blue_treasure_chest'),
  ('green_basket', 'green_bathtub'),
  ('green_bucket', 'red_crate'),
  ('blue_container', 'blue_container'),
  ('red_treasure_chest', 'red_treasure_chest'),
  ('blue_bucket', 'blue_bucket'),
  ('blue_drawer', 'blue_drawer'),
  ('bl

Done with 3 samples of 10 using the following indices splits: [0,10), [100,110), [200,210)
|Model|Accuracies|Avg Acc|
|--|--|--|
|GPT3.5| 0.7, 0.8, 0.4| 0.63|
|GPT4|0.7,0.6,0.4|0.57|

- GPT4 is saying there isn't enough info for some of them
- GPT4 more often says that there isn't enough information, whereas GPT3.5 just guesses wrong

Using random samples w/ seed 10, 15, 20:
|Model|Accuracies|Avg Acc|
|--|--|--|
|GPT3.5|0.8, 0.5, 0.8|0.7|
|GPT4| 0.7, 0.4, 0.7| 0.6|

- Higher accuracies here compared to before. It seems like certain stories trip up the model more than others.
- Why does GPT4 do worse?

5-shot learning with random samples(same seeds as above)
|Model|Accuracies|Avg Acc|
|--|--|--|
|GPT3.5|1,0.9, 0.9 |0.93|
|GPT4|0.9,0.8, 0.9|0.87|

5-shot learning on 50 random samples:
GPT4 - 82 percent, 
GPT3.5 - 0.78 percent

Hmm